io/EptReader.cpp

```diff
@@ -0,0 +1,461 @@
+    m_queryBounds.clip(q);
+
+    log()->get(LogLevel::Debug) << "Query origin " <<
+        *m_queryOriginId << ": " << found["path"].asString() <<
+        std::endl;
+}
+
+QuickInfo EptReader::inspect()
+{
+    QuickInfo qi;
+
+    initialize();
+
+    qi.m_bounds = toBox3d(m_info->json()["boundsConforming"]);
+    qi.m_srs = m_info->srs();
+    qi.m_pointCount = m_info->numPoints();
+
+    for (Json::ArrayIndex i(0); i < m_info->schema().size(); ++i)
+    {
+        qi.m_dimNames.push_back(m_info->schema()[i]["name"].asString());
+    }
+
+    qi.m_valid = true;
+
+    return qi;
+}
+
+void EptReader::addDimensions(PointLayoutPtr layout)
+{
+    const Json::Value& schema(m_info->schema());
+    m_remoteLayout.reset(new FixedPointLayout());
+
+    // Register XYZ directly since they may appear as scaled int32 values in the
+    // schema.  Either way we want doubles.
+    layout->registerDim(Dimension::Id::X);
+    layout->registerDim(Dimension::Id::Y);
+    layout->registerDim(Dimension::Id::Z);
+
+    for (Json::ArrayIndex i(0); i < schema.size(); ++i)
+    {
+        const Json::Value& dim(schema[i]);
+        const std::string name(dim["name"].asString());
+        log()->get(LogLevel::Debug) << "Registering dim " << name << std::endl;
+        const Dimension::Type type(Dimension::type(dim["type"].asString()));
+        layout->registerOrAssignDim(name, type);
+        m_remoteLayout->registerOrAssignDim(name, type);
+    }
+
+    m_remoteLayout->finalize();
+}
+
+void EptReader::ready(PointTableRef table)
+{
+    // Determine all the keys that overlap the queried area by traversing the
+    // EPT hierarchy (see https://git.io/fAiuR).  Because this may require
+    // fetching lots of JSON files, it'll run in our thread pool.
+    const Key key(m_info->bounds());
+    const Json::Value hier(parse(m_ep->get("h/" + key.toString() + ".json")));
+    overlaps(hier, key);
+    m_pool->await();
+
+    log()->get(LogLevel::Debug) << "Overlap nodes: " << m_overlapKeys.size() <<
+        std::endl;
+    log()->get(LogLevel::Debug) << "Overlap points: " << m_overlapPoints <<
+        std::endl;
+
+    if (m_overlapPoints > 1e8)
+    {
+        log()->get(LogLevel::Warning) << m_overlapPoints <<
+            " will be downloaded" << std::endl;
+    }
+
+    using D = Dimension::Id;
+
+    // Since we might need to revert a scale/offset for XYZ if the dataType is
+    // binary, they'll be handled separately.
+    m_dimTypes = table.layout()->dimTypes();
+    m_dimTypes.erase(std::remove_if(
+                m_dimTypes.begin(),
+                m_dimTypes.end(),
+                [](const pdal::DimType& d)
+                {
+                    return d.m_id == D::X || d.m_id == D::Y || d.m_id == D::Z;
+                }));
+}
+
+void EptReader::overlaps(const Json::Value& hier, const Key& key)
+{
+    if (!key.b.overlaps(m_queryBounds)) return;
+    const uint64_t np(hier[key.toString()].asUInt64());
```

Comment:
The data in question comes from the EPT format which shall not have anything but a positive integer here, or the value may not exist if we're checking one-past a leaf node, in which case it's a `Json::nullValue` so `0` is obtained.  Are you proposing a try-catch here (and potentially rethrowing a `pdal_error`)?

Reply:
I didn't know the best way to resolve.  If the JSON data you're parsing is always generated programmatically, then it doesn't much matter.  Otherwise, I'd just be concerned that you'd get some strange JSON exception that would be hard to understand.